designs/2019-esm-compatibilty/README.md

```diff
@@ -0,0 +1,195 @@
+
+### Story 2 - Node Compatiblity
+
+ESM support rolls out (as a non-experimental feature) w/ Node 13. This includes: the ability to define a package as a ESM; and backward interoperability w/ CJS in ES modules by using the `.cjs` file extension.
+
+This change *only* impacts users who define their package as an ES module. Without Node 13+ that isn't possible.
+
+Existing packages that use the default (ie CommonJS) will be unaffected.
+
+## Alternatives
+
+### Alternative 1 - Suppress the error
+
+**[PR#12333](https://github.com/eslint/eslint/pull/12333)**
+
+There is some debate among the node/modules group surrounding whether the error that causes this issue should be removed altogether.
+
+The fencing exists to prevent users from creating dual-mode (ie both ESM/CJS) packages as interoperability between the two formats has issues. In the case of ESLint, loading a simple config file from a user-defined package should not cause any issues or side-effecs.
+
+Eating the exception is viable solution. Although, this may become dead code if the error is removed from Node's core.
+
+### Alternative 2 - Support dynamic `import()`
+
+Instead of using 'import-fresh' to require the config, import it in a cross-format-compatible way using dynamic `import()`.
+
+There has been and will continue to be a lot of talk about using this approach as the CJS/ESM interop story is fleshed out.
+
+For now it presents too many unknowns and -- even if it didn't -- adding `import()` to ESLint's core would force Node 13+ as a hard constraint.
+
+## Open Questions
+
+<!--
+    This section is optional, but is suggested for a first draft.
+
+    What parts of this proposal are you unclear about? What do you
+    need to know before you can finalize this RFC?
+
+    List the questions that you'd like reviewers to focus on. When
+    you've received the answers and updated the design to reflect them, 
+    you can remove this section.
+-->
+
+## Help Needed
+
+<!--
+    This section is optional.
+
+    Are you able to implement this RFC on your own? If not, what kind
+    of help would you need from the team?
+-->
+
+## Frequently Asked Questions
+
+> Why doesn't compatibility w/ ESM-based packages 'just work'?
+
+ESLint reaches across the package boundary to retrieve the user-defined configuration. If the consumer package is ESM-based, the `.js` config file will be seen as an ES module, and trying to `require()` it will throw an error.
+
+> What does this interop issue affect?
+
+Only ESM-based packages (ie `"type": "module"`).
+
+> What about 3rd-party linting tools that provide their own built-in ruleset for ESLint (ex [StandardJS](https://standardjs.com/))?
+
+They aren't impacted by this change. 3rd-party modules define both their code and the ESLint ruleset used within the same package boundary.
```

Comment:
This might need some fleshing out.

The issue occurs when ESLint is doing the `require()` of a configuration (whether a directly-supplied configuration, or an extended configuration specified by `extends:` in configuration).

If a 3rd-party module provides a configuration object to ESLint, and this configuration object doesn't use `extends`, then 3rd-party modules wouldn't have issues because they handle the config resolution within their own package. However, if the 3rd-party module includes a configuration name/path that ESLint must resolve, then the error would occur, if my understanding is correct.

Is my understanding correct here?

Reply:
Maybe to clarify we could add "Aren't impacted as long as they don't explicitly add `type: module` to their own `package.json`."?

> However, if the 3rd-party module includes a configuration name/path that ESLint must resolve, then the error would occur, if my understanding is correct.

I think the error is relative to the file being required and independent of which callsite is calling `require`. So these two are equivalent:

```js
// in the tool, do own require
module.exports = require('some/other/config');

// in the tool, letting eslint do the require
module.exports = {
  extends: [require.resolve('some/other/config')],
};
```

In both cases, if `some/other/config` is inside of `type:module` and is a `.js` file, it will throw sooner or later. But since the code in those config packages should usually be targeted at ESLint and ESLint doesn't support async (ES module) configuration, it should be safe to assume that people won't put `type:module` into them.[1]

[1] This is different for `.eslintrc` files that may be in projects that are otherwise exclusively ES module code.

Reply:
> Maybe to clarify we could add "Aren't impacted as long as they don't explicitly add type: module to their own package.json."?

In Revision 2, I removed a lot of the unnecessary minutiae and made this abundantly clear.